### PR TITLE
do not cache large blob history event

### PR DESCRIPTION
**Pull request by yujieli-temporal**
**What changed?**

Allow lru cache size check against with CacheSize() method.

**Why?**
We want to limit the history event cache by BytesSize.

**How did you test it?**
unittest

**Potential risks**

**Is hotfix candidate?**

### CODE REVIEW

File: service/history/workflow/mutable_state_impl.go [hunk 1]
```diff
@@ -1184,6 +1184,12 @@ func (ms *MutableStateImpl) DeletePendingSignal(
 func (ms *MutableStateImpl) writeEventToCache(
 	event *historypb.HistoryEvent,
 ) {
+
+	// skip adding to cache if event is too large
+	if event.Size() > ms.config.HistoryCacheEventMaxSize() {
+		return
+	}
```
Comment (yycptt) [inline service/history/workflow/mutable_state_impl.go:1191]: I think this check should be done by the eventsCache impl? Mutable state may not be the only caller of PutEvent (in the future)?

**Comment by wxing1292** [comment]
take a look at https://github.com/temporalio/temporal/blob/master/common/cache/lru.go
1. create a duplicate of ^ and use cache size as cache eviction crateria  
2. modify the ^ implement, so ^ can support both size & count

File: common/cache/lru.go [hunk 2]
```diff
@@ -41,13 +45,15 @@ var (
...
-		maxSize  int
-		ttl      time.Duration
-		pin      bool
-		clock    clockwork.Clock
+		mut           sync.Mutex
+		byAccess      *list.List
+		byKey         map[interface{}]*list.Element
+		maxSize       int
+		maxTotalBytes int
+		totalBytes    int
```
Comment (wxing1292) [inline common/cache/lru.go:53]: can we use one unit?
this unit can be user defined, either size or count

Comment (yujieli-temporal) [reply common/cache/lru.go:53]: This change is on top of existing count base cache. Is up to we prefer allows the cache has 2 limits at same time or single limit. @yycptt do you have a preference?

File: common/cache/lru.go [hunk 3]
```diff
@@ -62,6 +68,10 @@ type (
 		value      interface{}
 		refCount   int
 	}
+
+	hasSizeMethod interface {
+		Size() int
+	}
```
Comment (wxing1292) [inline common/cache/lru.go:74]: maybe call this "TokenGetter" ?

e.g. say the cache has max token X
each item can consume Y token

File: common/cache/lru.go [hunk 4]
```diff
@@ -62,6 +63,10 @@ type (
 		value      interface{}
 		refCount   int
 	}
+
+	cacheSizeGetter interface {
+		CacheSize() int
+	}
```
Comment (wxing1292) [inline common/cache/lru.go:69]: i would suggest 

```go
itemSizeGetter interface {
	ItemSize() int
}
```

File: common/cache/lru.go [hunk 5]
```diff
@@ -300,20 +306,22 @@ func (c *lru) putInternal(key interface{}, value interface{}, allowUpdate bool)
 		entry.createTime = c.clock.Now().UTC()
 	}
 
-	if len(c.byKey) >= c.maxSize {
+	if c.currSize >= c.maxSize {
 		c.evictOnceInternal()
 	}
```
Comment (wxing1292) [inline common/cache/lru.go:311]: `if` should be changed to `for` loop?
e.g. what if a newly added item force evict 2 old items?

Comment (wxing1292) [reply common/cache/lru.go:311]: i guess at the very beginning of putting an item into cache
1. check if item size >= cache max size (or max allowed item size) -> if so error out earlier
2. if after adding the item, if cache size exceed limit, loop evicting old items until cache size <= threshold

**Review by wxing1292 — APPROVED** [review]

File: common/cache/lru.go (hunk 4 above)
Comment (yujieli-temporal) [reply common/cache/lru.go:71]: I think CacheSize() is more intuitive here to claim this method is for evaluate cache size.

Comment (yycptt) [reply common/cache/lru.go:71]: I think this interface should be exported, defined in `cache.go` and properly documented.

File: common/cache/lru.go
Comment (yycptt) [inline common/cache/lru.go:1]: It's a bit confusing that the `Size()` method of Cache still returns the count not the size.

File: common/cache/lru.go [hunk 6]
```diff
@@ -300,20 +313,22 @@ func (c *lru) putInternal(key interface{}, value interface{}, allowUpdate bool)
...
-	if len(c.byKey) >= c.maxSize {
+	for c.currSize >= c.maxSize {
 		c.evictOnceInternal()
 	}
-	if len(c.byKey) >= c.maxSize {
+	if c.currSize >= c.maxSize {
 		return nil, ErrCacheFull
 	}
 
 	element := c.byAccess.PushFront(entry)
```
Comment (yycptt) [inline common/cache/lru.go:323]: item size can be stored in `entryImpl` avoid computing the size multiple times.

File: common/cache/lru.go [hunk 7]
```diff
@@ -300,20 +313,22 @@ func (c *lru) putInternal(key interface{}, value interface{}, allowUpdate bool)
...
 	c.byKey[key] = element
+	c.currSize += getSize(value)
```
Comment (yycptt) [inline common/cache/lru.go:325]: c.currSize can exceed c.maxSize?

File: service/history/events/cache.go [hunk 8]
```diff
@@ -65,6 +65,14 @@ type (
 		metricsHandler metrics.Handler
 		shardID        int32
 	}
+
+	historyEventItem interface {
+		CacheSize() int
+	}
```
Comment (yycptt) [inline service/history/events/cache.go:71]: I don't think the interface needs to be defined again here

File: service/history/events/cache.go [hunk 9]
```diff
@@ -210,3 +218,21 @@ func (e *CacheImpl) getHistoryEventFromStore(
...
 	return nil, errEventNotFoundInBatch
 }
+
+func (e *CacheImpl) put(key EventKey, event *historypb.HistoryEvent) interface{} {
+	return e.Put(key, NewHistoryEventCacheItem(event))
+}
+
+var _ historyEventItem = (*historyEventItemImpl)(nil)
+
+func NewHistoryEventCacheItem(
```
Comment (yycptt) [inline service/history/events/cache.go:228]: this constructor should not be exported I think. it's the impl detail of history events cache

File: common/cache/lru.go
Comment (yujieli-temporal) [reply common/cache/lru.go:1]: sorry, this is a miss, will change.

File: common/cache/lru.go (hunk 7 above)
Comment (yujieli-temporal) [reply common/cache/lru.go:325]: It will, but items will be evict after next put until currSize is under maxSize.

File: common/cache/lru.go [hunk 10]
```diff
@@ -247,12 +257,12 @@ func (c *lru) Release(key interface{}) {
 	entry.refCount--
 }
 
-// Size returns the number of entries currently in the lru, useful if cache is not full
+// Size returns the current size of the lru, useful if cache is not full
```
Comment (yycptt) [inline common/cache/lru.go:260]: plz update the comment for the Size() method in `cache.go` as well. Plz also mention that the size calculation depends on if the item implements the SizeGetter interface

File: common/cache/size_getter.go
Comment (yycptt) [inline common/cache/size_getter.go:1]: nit: I think this is part of the cache interface contract actually and should be moved to that file, but I don't have strong opinion.

File: common/cache/lru.go (hunk 7 above)
Comment (yycptt) [reply common/cache/lru.go]: This essentially change the maxSize from a hard limit to a soft limit. Should be ok but I don't see a reason for breaking that if it's easy to keep it it as a hard limit. On L320 the condition can be `c.currSize + size > c.maxSize`, then we get a hard limit?

File: common/cache/lru.go [hunk 11]
```diff
@@ -300,20 +317,22 @@ func (c *lru) putInternal(key interface{}, value interface{}, allowUpdate bool)
 		entry.createTime = c.clock.Now().UTC()
 	}
 
-	if len(c.byKey) >= c.maxSize {
+	if c.currSize >= c.maxSize {
```
Comment (yycptt) [inline common/cache/lru.go:320]: I remember this is a `for` loop in the previous version

Comment (yujieli-temporal) [reply common/cache/lru.go:320]: I changed it back, from current implementation the LRU will not pop out any item if the reference is not 0. so the for loop here will introduce an infinite loop.

**Review by wxing1292 — APPROVED** [review]
LGTM

File: common/cache/lru.go [hunk 12]
```diff
@@ -290,6 +302,7 @@ func (c *lru) putInternal(key interface{}, value interface{}, allowUpdate bool)
 	entry := &entryImpl{
 		key:   key,
 		value: value,
+		size:  getSize(value),
```
Comment (yycptt) [inline common/cache/lru.go:305]: shall we still do the check of size v.s. cache max size and return error early if item size is too big.

**Review by yycptt — APPROVED** [review]

File: common/cache/lru.go (hunk 12 above)
Comment (yujieli-temporal) [reply common/cache/lru.go]: sure, I move the size validation on the top before entry created.

Comment (yycptt) [reply common/cache/lru.go]: I guess I mean we don't have to evict any entry if new item size is too big.

Comment (yujieli-temporal) [reply common/cache/lru.go]: I see, thanks for clarify. I don't feel strong here, I will make the change to do the check before.

Comment (yujieli-temporal) [reply common/cache/lru.go]: actually, run the loop twice will make the logic looks duplicate and complex. do you feel strong about check before eviction?

File: common/cache/lru.go [hunk 13]
```diff
@@ -261,9 +274,22 @@ func (c *lru) putInternal(key interface{}, value interface{}, allowUpdate bool)
 	if c.maxSize == 0 {
 		return nil, nil
 	}
+	entrySize := getSize(value)
+	if entrySize >= c.maxSize {
```
Comment (yycptt) [inline common/cache/lru.go:278]:
```suggestion
	if entrySize > c.maxSize {
```

**Review by yycptt — APPROVED** [review]

**Pull request merged**